### PR TITLE
[experimental] Add OpenReward Standard environment adapter

### PR DESCRIPTION
**Pull request by adithya-s-k**
# What does this PR do?

Adds `trl.experimental.openreward` — an adapter that plugs any environment speaking the [**Open Reward Standard (ORS)**](https://openrewardstandard.io) into any TRL trainer that accepts an `environment_factory` input (e.g. `GRPOTrainer`, `AsyncGRPOTrainer`, `RLOOTrainer`, …). ORS is an open HTTP/SSE protocol for RL environments; the same adapter works for envs hosted on the [**OpenReward platform**](https://openreward.ai), self-hosted on any container service, or running locally on `localhost` for development. One string identifier (a catalog name like `Eigent/SETA` or a URL like `http://localhost:8000`) is enough to wire all three TRL slots: `.dataset`, `.factory`, `.reward_func`.

Discussed in #5695.

```python
from trl import GRPOConfig, GRPOTrainer
from trl.experimental.openreward import OpenRewardEnv

env = OpenRewardEnv("Eigent/SETA")  # or "http://localhost:8000"

trainer = GRPOTrainer(
    model="Qwen/Qwen3-4B",
    args=GRPOConfig(...),
    train_dataset=env.dataset,
    environment_factory=env.factory,
    reward_funcs=env.reward_func,
)
```

## What this PR adds

| Area | What |
|---|---|
| Adapter | `trl/experimental/openreward/` — `OpenRewardEnv` (spec) + `_RolloutEnvironment` (internal, per-rollout) |
| Example | `examples/scripts/openreward/seta.py` — full GRPO script targeting the SETA environment, runnable in colocate or server vLLM mode |
| Optional dep | `trl[openreward]` extra in `pyproject.toml` (lazy import — non-users pay nothing; mirrors `trl[peft]`, `trl[vllm]`) |

5 files, +774 LOC including docstrings.

## Design choices

- **One public class.** `OpenRewardEnv` is the spec; the per-rollout `_RolloutEnvironment` is internal (leading underscore, not in `__all__`).
- **SDK-backed transport.** Wraps the official [`openreward` Python SDK](https://pypi.org/project/openreward/) rather than re-implementing the protocol — keeps the adapter small and tracks protocol changes upstream.
- **Dynamic tool binding.** At construction, `OpenRewardEnv` fetches the env's tool schemas (over the ORS `/tools` endpoint) and generates one Python method per tool with a proper typed signature + docstring. No per-env wrapper code — `inspect.getmembers(env, ismethod)` and `transformers.utils.get_json_schema` work out of the box.
- **Dataset autoderivation.** `env.dataset` resolves the env's task list via the ORS task endpoints (`/num_tasks`, `/task`, `/task_range`, `/splits`) to build a `datasets.Dataset` with `task_index` + per-task metadata columns.

## Why now

The single example here (`seta.py`) is meant as a starting point — the adapter itself is environment-agnostic, so any current or future ORS env works against the same code path. As more environments land on the OpenReward catalog, additional examples can be added incrementally without touching the adapter.

## Open follow-ups (post-merge, not in this PR)

- Docs page at `docs/source/openreward.md` (toctree near `openenv.md`).
- Tests at `tests/experimental/test_openreward.py` with an in-process FastAPI fake to keep CI off the network.
- Graduate `experimental/openreward` → `trl.openreward` after one release cycle if no breaking ORS changes.
- 15-min ORS session keepalive `ping` thread for very long episodes.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? → #5695
- [ ] Did you make sure to update the documentation with your changes? *(deferred — see follow-ups above; happy to add the docs page in this PR if reviewers prefer)*
- [ ] Did you write any new necessary tests? *(deferred — same)*

## AI writing disclosure

- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.

---

> [!NOTE]
> **Medium Risk**
> Adds a new experimental environment integration that performs dynamic tool-method generation (via `exec`) and opens real HTTP/SSE sessions, so failures or schema quirks could affect environment-based training behavior. Risk is mitigated by being gated behind the `trl[openreward]` extra and covered by end-to-end tests against a local echo server.
> 
> **Overview**
> Adds **experimental OpenReward (ORS) environment support** via `trl.experimental.openreward.OpenRewardSpec`, which lazily derives a `train_dataset`, provides an `environment_factory` that creates per-rollout ORS sessions, and ships a default outcome-only `reward_funcs`.
> 
> Implements a per-rollout adapter that **binds ORS tools into typed Python methods** from JSON Schema at runtime and tracks rollout state (`reward`, `finished`, etc.) for reward computation.
> 
> Rounds out the integration with the `trl[openreward]` optional dependency, documentation + a SETA training example script, and new end-to-end tests that spin up a minimal local ORS echo server (plus a `require_openreward` test marker).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4e9baa0b6f53a0e08bca993cfc6041d8b40b420b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: examples/scripts/openreward/seta.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        max_completion_length=args.max_completion_length,
+        max_steps=args.max_steps,
+        max_tool_calling_iterations=args.max_tool_calling_iterations,
+        chat_template_kwargs={"enable_thinking": False},
+        log_completions=True,
+        bf16=True,
+        gradient_checkpointing=True,
+        gradient_checkpointing_kwargs={"use_reentrant": False},
+        beta=0.01,
+        loss_type="grpo",
```
Comment (qgallouedec) [inline examples/scripts/openreward/seta.py]:
```suggestion
```

- bf16 is true by default
- grad chkpt is enable by default
- `"use_reentrant"` is set to false by default 
- unless there a is a very good reason, I'd recommend deactivating the kl term (default)
- I think that the default "dapo" is usually a better choice

File: examples/scripts/openreward/seta.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        config_kwargs["output_dir"] = args.output_dir
+    if args.vllm_mode == "colocate":
+        config_kwargs["vllm_gpu_memory_utilization"] = args.vllm_gpu_memory_utilization
+    else:
+        config_kwargs["vllm_server_base_url"] = args.vllm_server_base_url
+
+    trainer = GRPOTrainer(
+        model=args.model,
+        args=GRPOConfig(**config_kwargs),
+        train_dataset=env.dataset,
```
Comment (qgallouedec) [inline examples/scripts/openreward/seta.py]: It's not for this PR, but it would simplify the API if we allow `train_dataset` to be `None` when an environment is provided. We would have something like 


```python
trainer = GRPOTrainer(
    model=args.model,
    args=GRPOConfig(**config_kwargs),
    environment_factory=env.factory,
    reward_funcs=env.reward_func,
)
```

or even for reward:

```python
trainer = GRPOTrainer(
    model=args.model,
    args=GRPOConfig(**config_kwargs),
    environment_factory=env.factory,
)
```

File: trl/experimental/openreward/_spec.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        time, so each rollout has an isolated ORS session. Reuses the
+        spec's already-discovered SDK env + tool specs to skip per-env
+        HTTP at trainer init.
+        """
+        # Pre-fetch tool specs once at the spec level.
+        env = self._sdk_env
+        tool_specs = env.list_tools()
+        client = env.client if hasattr(env, "client") else self._sdk_client
+
+        kwargs: dict[str, Any] = {
```
Comment (qgallouedec) [inline trl/experimental/openreward/_spec.py]:
```suggestion
        kwargs = {
```

turbo nit

File: trl/experimental/openreward/_spec.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    """Default reward function: last non-null reward in each rollout's trajectory.
+
+    Suitable for sparse-outcome envs (e.g. SETA, where only
+    `submit_solution` returns a non-null reward). Override by passing a
+    different callable to ``reward_funcs=``.
+    """
+    return [env.reward for env in environments]
+
+
+class OpenRewardEnv:
```
Comment (qgallouedec) [inline trl/experimental/openreward/_spec.py]: I think the name is a bit misleading. It seems like you get an "Env" i.e. "Environment", which is in my understanding an interface that you can step, reset etc. Whereas here, you get something a bit different, some kind of environment factory with additional features like dataset and reward. I'm not entirely sure what would be a better name, but "Env" sounds not accurate

**Review by qgallouedec — COMMENTED** [review]
This looks good! 

I think it misses a dedicated section in the doc, possibly converging how to create a minimal OpenReward environment (on the hub or locally), and how to use this to train with GRPO.

It also misses tests. Is it feasible to create a dummy environment for test? Either on the hub in `til-internal-testing`, or deployed locally

**Comment by qgallouedec** [comment]
>  (deferred — see follow-ups above; happy to add the docs page in this PR if reviewers prefer)

Yes, I'd prefer, it doesn't need to be complete, but at least have some basic guidance, that we will polish in the future. Same for testing

File: trl/experimental/openreward/_spec.py (hunk 4 above)
Comment (adithya-s-k) [reply trl/experimental/openreward/_spec.py]: shall we call it `OpenRewardSpec`

**Comment by adithya-s-k** [comment]
> > (deferred — see follow-ups above; happy to add the docs page in this PR if reviewers prefer)
> 
> Yes, I'd prefer, it doesn't need to be complete, but at least have some basic guidance, that we will polish in the future. Same for testing

I will add documentation and test cases along with that, for the test can can i emulate the Openreward server inprocess with fastapi for testing , or do you want me to push an environment to spaces and use that to test things out?

**Comment by qgallouedec** [comment]
both options sound valid, your call

File: trl/experimental/openreward/_spec.py (hunk 4 above)
Comment (qgallouedec) [reply trl/experimental/openreward/_spec.py]: `OpenRewardSpec` could work

**Comment by adithya-s-k** [comment]
Pushed b7861f16 with the renames from your review:

- `OpenRewardEnv` → `OpenRewardSpec` (the "Env" naming was misleading — what we have is really a spec/factory)
- `.dataset` / `.factory` / `.reward_func` → `.train_dataset` / `.environment_factory` / `.reward_funcs`. Each property now maps 1:1 to the matching `GRPOTrainer` kwarg, so the call site reads:

```python
spec = OpenRewardSpec("Eigent/SETA")
trainer = GRPOTrainer(
    train_dataset=spec.train_dataset,
    environment_factory=spec.environment_factory,
    reward_funcs=spec.reward_funcs,
    ...
)
```

- Dropped the redundant defaults from `seta.py` (bf16, grad_chkpt, use_reentrant, beta, loss_type="grpo") since they all match the `GRPOConfig` defaults already.
- Turbo nit applied.

@qgallouedec on the `train_dataset=None` / `reward_funcs=None` simplification — happy to do that as a follow-up PR. The property renames above already set it up since they match the trainer kwargs, so once the trainer auto-derives them from the env, the call site collapses to:

```python
spec = OpenRewardSpec("Eigent/SETA")
trainer = GRPOTrainer(
    model="Qwen/Qwen3-4B",
    args=GRPOConfig(...),
    environment_factory=spec.environment_factory,
)
```

Will add the docs and tests soon.

File: pyproject.toml
Comment (qgallouedec) [inline pyproject.toml:112]:
```suggestion
    # openreward
    "openreward>=0.1.109",
    # peft
```

**Comment by qgallouedec** [comment]
nice first iteration! make sure to run `make precommit` before merging

File: docs/source/openreward.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+trainer = GRPOTrainer(
+    ...,
+    reward_funcs=my_reward,
+)
+```
+
+The per-rollout adapter exposes the running state TRL needs — `env.reward`, `env.rewards`, `env.metadata`, `env.finished`, `env.last_output` — for arbitrary post-hoc reward shaping.
+
+## Limitations
```
Comment (qgallouedec) [inline docs/source/openreward.md:196]:
```suggestion

## OpenRewardSpec

[[autodoc]] trl.experimental.openreward.OpenRewardSpec

## Limitations
```

**Review by qgallouedec — APPROVED** [review]
Ok, I think this looks good!

File: trl/experimental/openreward/environment.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+
+    # ── helpers ──────────────────────────────────────────────────────
+
+    def _teardown_session(self) -> None:
+        if self._session is None:
+            return
+        try:
+            # The SDK's Session context-manager exit deletes both the env
+            # state and the sid. We look up the original CM from the SDK.
+            self._session.__exit__(None, None, None) if hasattr(self._session, "__exit__") else None
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/openreward/environment.py]: fair, this could be improved

**Comment by qgallouedec** [comment]
@codex review

File: trl/experimental/openreward/environment.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adithya-s-k) [reply trl/experimental/openreward/environment.py]: reformatted accodingly

File: examples/scripts/openreward/seta.py
Comment (sergiopaniego) [inline examples/scripts/openreward/seta.py:1]: The example should be linked in `example_overview.md`.

File: trl/experimental/openreward/environment.py [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+            Name to look up on the server. Defaults to the canonical name parsed from `name`/`base_url`.
+        split (`str`, *optional*, defaults to `"train"`):
+            Split passed to ``env.session(split=, index=)``.
+        api_key (`str`, *optional*):
+            Override for ``OPENREWARD_API_KEY``. Only used with `name=`.
+        secrets (`dict[str, str]`, *optional*):
+            Per-session secrets. Forwarded to ``env.session(secrets=)``; the SDK encodes them and adds platform-domain
+            entries.
+        timeout (`float`, *optional*):
+            Reserved for future use. The SDK manages its own timeouts.
```
Comment (sergiopaniego) [inline trl/experimental/openreward/environment.py:130]: `Reserved for future use.`. If it's something that is not currently used, could be removed, no?

**Review by sergiopaniego — COMMENTED** [review]
Great work @adithya-s-k! Leaving a couple of ideas 😄
